File: web/src/api/llm.js
// Import Axios for making HTTP requests to the backend API
import axios from "axios";

// Send a prompt to the LLM query endpoint and return the generated response
export async function queryLLM(input, options = {}) {
  const payload = {
    input,                                       // User input / prompt
    max_tokens: options.max_tokens || 100,       // Default to 100 tokens
    temperature: options.temperature || 0.7,     // Default temperature
  };

  // POST the query to the Synthara API
  const res = await axios.post("/api/llm/query", payload);

  // Return the API response
  return res.data;
}

// Send training texts to fine-tune the model via the backend API
export async function trainLLM(texts) {
  const res = await axios.post("/api/llm/train", { texts });
  return res.data;
}

File: backend/api/routes/llm.py
# Import FastAPI router and request context
from fastapi import APIRouter, Request

# Import Pydantic for validating incoming payloads
from pydantic import BaseModel

# Import global engine instance (provides access to model and DB)
from backend.core.engine import engine

# PyTorch for tensor operations and inference control
import torch

# Create a FastAPI router for LLM-related endpoints
router = APIRouter()

# Define expected input payload for querying the model
class QueryPayload(BaseModel):
    input: str                  # User input prompt
    max_tokens: int = 100       # Max tokens to generate in response
    temperature: float = 0.7    # Sampling temperature for variability

# POST endpoint to query the language model
@router.post("/api/llm/query")
async def query_llm(payload: QueryPayload):
    # Retrieve model and tokenizer from the Synthara engine
    llm = engine.get_model()
    model = llm["model"]
    tokenizer = llm["tokenizer"]

    # Tokenize user input and move it to the model's device (e.g. CPU or GPU)
    inputs = tokenizer(payload.input, return_tensors="pt").to(model.device)

    # Generate a response without computing gradients (inference-only)
    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=payload.max_tokens,
            temperature=payload.temperature,
            do_sample=True
        )

    # Decode the model output into human-readable text
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    return {"response": response}
